crates/ra_mbe/src/mbe_expander.rs

```diff
@@ -8,19 +8,33 @@ mod transcriber;
 use ra_syntax::SmolStr;
 use rustc_hash::FxHashMap;
 
-use crate::ExpandError;
+use crate::{ExpandResult, ExpandError};
 
 pub(crate) fn expand(
     rules: &crate::MacroRules,
     input: &tt::Subtree,
-) -> Result<tt::Subtree, ExpandError> {
-    rules.rules.iter().find_map(|it| expand_rule(it, input).ok()).ok_or(ExpandError::NoMatchingRule)
+) -> ExpandResult<tt::Subtree> {
+    let (mut result, mut left_over, mut err) = (tt::Subtree::default(), usize::max_value(), Some(ExpandError::NoMatchingRule));
+    for rule in &rules.rules {
+        let ((res, left), e) = expand_rule(rule, input);
```

Comment:
Another option is uplift `ExpandCtx` and collect these errors there, something likes:
```rust
let ctx = ExpandCtx::default();
if let Err(err) = expand_rule(rule, input, &mut ctx) {
   if ctx.left_over < left {
      ...
   }
}

```